R/call.R

```diff
@@ -651,12 +661,129 @@ call_standardise <- function(call, env = caller_env()) {
     fn <- eval_bare(node_car(expr), env)
   }
 
-  if (is_primitive(fn)) {
-    call
+  matched <- call_match(
+    expr,
+    fn,
+    defaults = defaults,
+    dots_env = dots_env
+  )
+  set_expr(call, matched)
+}
+
+#' Match supplied arguments to function definition
+#'
+#' @description
+#' `call_match()` is like [match.call()] with these differences:
```

Comment:
Would it make sense to document `call_match()` and `call_standardise()` together? I don't quite understand what the difference is, and I'm kind of surprised that `call_standardise()` calls `call_match()`, rather than the other way around

Reply:
I thought it'd be confusing to document them together since they are so close. I think the difference is that `call_standardise()` is supposed to be a little more user friendly because it's intended to work with a single argument (a call) rather than two or three (call, definition, dots environment).

Personally I find its UI a bit dangerous regarding the scoping, it's easy to make logical mistakes when not careful. It's inherited from lazyeval where the assumption was that most calls would be wrapped in formulas, thus carrying an environment. With rlang it seemed like the same principle would apply with quosures. In this context, supplying a function definition is superfluous because it can be inferred from the environment of the formula or quosure. However time has shown that we rarely work with wrapped expressions but with bare expressions, and then not requiring an `env` or `fn` argument is not ideal.

That's why `call_match()` takes a different approach.

- Unlike `match.call()`, it either infers everything from the call stack or nothing.
- Unlike `call_standardise()`, it forces you to supply a function definition.

Maybe we should supersede `call_standardise()`?

Reply:
Ah ok, that makes sense to me. You can make the call on whether or not to supersede `call_standardise()`, but if we don't we need to make sure to cross-link the functions.